### ISSUE TITLE
sclang: steady memory leak when server is booted

### ISSUE THREAD

**Issue by snappizz** [supercollider/supercollider#4268]
Environment
-----------

* Your SuperCollider version: 3.10.1
* Your operating system and version: Ubuntu 16.04

Steps to reproduce
-----------------------------

here's a python script to monitor changes in RSS (use python 3.7+):

```python
import subprocess
import time

def get_rss():
    process = subprocess.run("ps -eo rss,comm | grep sclang", capture_output=True, shell=True, encoding="utf-8")
    res = process.stdout.strip()
    if not res:
        return 0
    rss = int(res.split()[0])
    return rss

rss = None
seconds = 0
while True:
    time.sleep(1)
    new_rss = get_rss()
    if rss != None and new_rss != rss:
        change = new_rss - rss
        print("RSS = {}KB, change = {}KB over {}s".format(new_rss, change, seconds))
        seconds = 0
    else:
        print(".")
    rss = new_rss
    seconds += 1
```

make a supercollider file (`server.scd`) whose entire contents are `Server.default.boot`. run `sclang server.scd` with the python script running, and watch the python output. it may take around 30 seconds for the serious memory leakage to occur. the leaks i see are on the order of 250 KB at 4-6 second intervals.

Comments
---------

memory is leaking in spurts about every few seconds, so it's pretty clearly caused (directly or indirectly) by server status messages. this is confirmed by stopping and starting the watcher:

```supercollider
Server.default.statusWatcher.stopAliveThread // leaking stops
Server.default.statusWatcher.startAliveThread // leaking starts again
```

**Comment by snappizz** [comment]
having issues with valgrind (first [this boost issue](https://svn.boost.org/trac10/ticket/10005), then an inexplicable crash), so i've been probing this with my terrible `ps` hack job.

i have traced the bug to the `OSCFunc` created in `ServerStatusWatcher:addWatcher`. if it's never created, or never triggered by a `/status`->`/status.reply` roundtrip, then the leak doesn't happen.

**Comment by dyfer** [comment]
I can (sadly) confirm this also happens on macOS 10.13 with SC 3.10.0
(does not happen with SC 3.9.3)

BTW ps command seem to behave slightly differently on a mac.
<details><summary>Here's the modified python script that works on macOS</summary>

```python
import subprocess
import time

def get_rss():
    process = subprocess.run("ps -eo rss,comm | grep sclang", capture_output=True, shell=True, encoding="utf-8")
    rss = int(process.stdout.strip().split()[0])
    return rss

rss = None
seconds = 0
while True:
    time.sleep(1)
    new_rss = get_rss()
    if rss != None and new_rss != rss:
        change = new_rss - rss
        print("RSS = {}KB, change = {}KB over {}s".format(new_rss, change, seconds))
        seconds = 0
    rss = new_rss
    seconds += 1
```

</details>

**Comment by snappizz** [comment]
bisect complete:

```shell
git bisect start
# bad: [e567b992bd11bded747b820be576e628c7c08035] Merge pull request #4257 from snappizz/topic/3.10.1
git bisect bad e567b992bd11bded747b820be576e628c7c08035
# good: [29cb826639a33b4986654750c9df70de09c9fa30] qcscopeshm: rewrite inner paint1D loop
git bisect good 29cb826639a33b4986654750c9df70de09c9fa30
# bad: [213e65756bd3aa0ba55ffcb7f45e09a43cbc8de1] Merge pull request #3918 from telephon/topic/switch-plot-local-remote
git bisect bad 213e65756bd3aa0ba55ffcb7f45e09a43cbc8de1
# bad: [52a0296fbb89bc09137dee168b619555a27f09c7] Merge pull request #2504 from supercollider/topic/menu-support
git bisect bad 52a0296fbb89bc09137dee168b619555a27f09c7
# bad: [17b9457f5bceebe7baaa63e8f4189454c055095f] sclang: fix possible UI access on wrong thread
git bisect bad 17b9457f5bceebe7baaa63e8f4189454c055095f
# good: [3b56c5fabfeb839d2e4d384177679cd369ed4e8c] Merge pull request #3767 from snappizz/topic/linuxreadme
git bisect good 3b56c5fabfeb839d2e4d384177679cd369ed4e8c
# good: [e40ef28cdc750be0c32953941650710adce2d45c] classlib: ToolButtonStyle
git bisect good e40ef28cdc750be0c32953941650710adce2d45c
# bad: [1c6b6159b1cbe1286ec7d4a267e94f33fbda637c] Merge branch 'topic/webengine' into topic/menu-support
git bisect bad 1c6b6159b1cbe1286ec7d4a267e94f33fbda637c
# bad: [cc4882a2c9db0e86a24c1e5f3e58691f1e61d782] Merge branch 'topic/menu-support' of https://github.com/supercollider/supercollider into topic/menu-support
git bisect bad cc4882a2c9db0e86a24c1e5f3e58691f1e61d782
# good: [964d1f860b322b5afb840c65e0ed3abd0866142e] sclang: Allow either QAction or index when inserting actions.
git bisect good 964d1f860b322b5afb840c65e0ed3abd0866142e
# skip: [c2b3ea3f2b4013d0e1b68b5c95b84e16c31de761] sclang: Ordered app menus.
git bisect skip c2b3ea3f2b4013d0e1b68b5c95b84e16c31de761
# bad: [7a56b269734fb08534b0e228421087f83d325fbe] classlib: Better MainMenu implementation
git bisect bad 7a56b269734fb08534b0e228421087f83d325fbe
# first bad commit: [7a56b269734fb08534b0e228421087f83d325fbe] classlib: Better MainMenu implementation
```

this is pretty bizarre -- i'd expect this to be some deep C++ thing but  7a56b269734fb08534b0e228421087f83d325fbe is actually a class library commit. interesting.

given the somewhat nondeterministic/subjective method for figuring out whether the memory leak has happened, the bisection process here isn't necessarily infallible, but i can verify that this commit experiences the leak and the previous commit does not.

**Comment by snappizz** [comment]
wow, the source of this leak is something i would never have guessed without git bisect! it's `MainMenu`. an unfortunate combination of three issues:

1. recreating and destroying an entire menu every time the `Server` object changes due to adding all `Server`s as dependants. the recreation/destruction is indirectly triggered by the server status watcher
2. a memory leak somewhere in that process of continual recreation and destruction
3. `MainMenu` is unconditionally created and takes up resources even on desktop environments where it's not going to be visible (its behavior only makes sense on macOS and ubuntu unity).

have not tested yet, but i think an already existing PR -- #3870 -- will at least address problem 1 and fix this leak for the time being.

cc @scztt

**Comment by snappizz** [comment]
plan for fixing this issue:

- complete and merge #3870 before 3.10.2. i have tested and it at least prevents the leak from happening.
- if there's a lower-level leak happening here, let's fix that up of course.
- add a new method, `MainMenu.platformHasNativeMenuBar`, which returns true (i have a branch for this going). make sure `MainMenu.initClass` is only called when this is the case.

**Comment by miczac** [comment]
Hi, out of curiosity I let SC run for quite a while and I of course could observe this issue.
However, how would you explain this?
```
mz$ ps -eo rss,comm | grep sclang
639448 /Users/mz/dev/sc3.git/3.10/build/Install/SuperCollider/SuperCollider.app/Contents/Resources/../MacOS/sclang
mz$ ps -eo rss,comm | grep sclang
875996 /Users/mz/dev/sc3.git/3.10/build/Install/SuperCollider/SuperCollider.app/Contents/Resources/../MacOS/sclang
mz$ ps -eo rss,comm | grep sclang
599308 /Users/mz/dev/sc3.git/3.10/build/Install/SuperCollider/SuperCollider.app/Contents/Resources/../MacOS/sclang
mz$ ps -eo rss,comm | grep sclang
618064 /Users/mz/dev/sc3.git/3.10/build/Install/SuperCollider/SuperCollider.app/Contents/Resources/../MacOS/sclang

```
I didn't run _ps_ on a regular basis but this  lines are in correct order and off a longer sequence of checks.
Why is this  drop in memory usage? 
This may be the reason why it didn't really occur problematic to me, although I tend to have SC running for days.
(that's on OS X 10.12.6)

**Comment by brianlheim** [comment]
@miczac sclang is a garbage collected language. if gc runs, memory will be freed

**Comment by miczac** [comment]
ok, thought so  but wasn't sure .. thanks!

**Comment by snappizz** [comment]
this issue affecting the majority of SuperCollider users been addressed in #3870 and #4285 -- closing in favor of #4288 for discussion of the core sclang memory leak.

**Comment by snappizz** [comment]
one small thing i want to add here -- part of what made this bug hard to find is the use of `addDependant`, which almost "hides" code by fragmenting it across different classes in a very non-obvious way. i think use of this functionality should be avoided to make code easier to follow.
